Can't run in readthedocs

Hi @minrk we have a jupyterbook project where we install fenics with conda.

[Two days ago it worked fine](https://readthedocs.org/projects/festim-vv-report/builds/24775144/) and today we started noticing [errors in the read the docs build](https://festim-vv-report--35.org.readthedocs.build/en/35/validation/plasma-driven-permeation/plasma-driven-permeation.html).

It seems to be a JIT problem:

```
------------------- Start compiler output ------------------------
/usr/bin/ld: cannot find /lib64/libpthread.so.0: No such file or directory
/usr/bin/ld: cannot find /usr/lib64/libpthread_nonshared.a: No such file or directory
collect2: error: ld returned 1 exit status
-------------------  End compiler output  ------------------------
Compilation failed! Sources, command, and errors have been written to: /home/docs/checkouts/readthedocs.org/user_builds/festim-vv-report/checkouts/35/report/validation/plasma-driven-permeation/jitfailure-dolfin_expression_a3e880d7e745db7ff656246e44892686
```

I see there have been changes recently to this repo, could this be related?

Thanks!

Full error message:

```
---------------------------------------------------------------------------
DijitsoError                              Traceback (most recent call last)
File ~/checkouts/readthedocs.org/user_builds/festim-vv-report/conda/35/lib/python3.11/site-packages/dolfin/jit/jit.py:165, in compile_class(cpp_data, mpi_comm)
    164 try:
--> 165     module, signature = dijitso_jit(cpp_data, module_name, params,
    166                                     generate=cpp_data['jit_generate'],
    167                                     mpi_comm=mpi_comm)
    168     submodule = dijitso.extract_factory_function(module, "create_" + module_name)()

File ~/checkouts/readthedocs.org/user_builds/festim-vv-report/conda/35/lib/python3.11/site-packages/dolfin/jit/jit.py:47, in mpi_jit_decorator.<locals>.mpi_jit(*args, **kwargs)
     46 if MPI.size(mpi_comm) == 1:
---> 47     return local_jit(*args, **kwargs)
     49 # Default status (0 == ok, 1 == fail)

File ~/checkouts/readthedocs.org/user_builds/festim-vv-report/conda/35/lib/python3.11/site-packages/dolfin/jit/jit.py:103, in dijitso_jit(*args, **kwargs)
    101 @mpi_jit_decorator
    102 def dijitso_jit(*args, **kwargs):
--> 103     return dijitso.jit(*args, **kwargs)

File ~/checkouts/readthedocs.org/user_builds/festim-vv-report/conda/35/lib/python3.11/site-packages/dijitso/jit.py:216, in jit(jitable, name, params, generate, send, receive, wait)
    214 if err_info:
    215     # TODO: Parse output to find error(s) for better error messages
--> 216     raise DijitsoError("Dijitso JIT compilation failed, see '%s' for details"
    217                        % err_info['fail_dir'], err_info)
    219 # Return built library and its signature

DijitsoError: Dijitso JIT compilation failed, see '/home/docs/checkouts/readthedocs.org/user_builds/festim-vv-report/checkouts/35/report/validation/plasma-driven-permeation/jitfailure-dolfin_expression_a3e880d7e745db7ff656246e44892686' for details

During handling of the above exception, another exception occurred:

RuntimeError                              Traceback (most recent call last)
Cell In[3], line 39
     19 my_model.boundary_conditions = [
     20     F.RecombinationFlux(
     21         Kr_0=1.0e-27 * (1.0 - 0.9999 * sp.exp(-6.0e-5 * F.t)),
   (...)
     26     F.RecombinationFlux(Kr_0=2.0e-31, E_Kr=0, surfaces=[2], order=2),
     27 ]
     29 flux = sp.Piecewise(
     30     (4.9e19, F.t < 5820),
     31     (0, F.t < 9060),
   (...)
     36     (0, True),
     37 )
     38 my_model.sources = [
---> 39     F.ImplantationFlux(flux=flux, imp_depth=12e-9, width=2.4e-9, volume=1)
     40 ]
     42 my_model.T = F.Temperature(500)  # ignored here
     44 left_flux = F.SurfaceFlux(surface=1, field="solute")

File ~/checkouts/readthedocs.org/user_builds/festim-vv-report/conda/35/lib/python3.11/site-packages/festim/sources/source_implantation_flux.py:42, in ImplantationFlux.__init__(self, flux, imp_depth, width, volume)
     36 distribution = (
     37     1
     38     / (self.width * (2 * np.pi) ** 0.5)
     39     * sp.exp(-0.5 * ((x - self.imp_depth) / self.width) ** 2)
     40 )
     41 value = self.flux * distribution
---> 42 super().__init__(value, volume, field="0")

File ~/checkouts/readthedocs.org/user_builds/festim-vv-report/conda/35/lib/python3.11/site-packages/festim/sources/source.py:31, in Source.__init__(self, value, volume, field)
     29     self.value = Constant(value)
     30 elif isinstance(value, sp.Expr):
---> 31     self.value = Expression(sp.printing.ccode(value), t=0, degree=2)
     32 elif isinstance(value, (Expression, UserExpression, Function)):
     33     self.value = value

File ~/checkouts/readthedocs.org/user_builds/festim-vv-report/conda/35/lib/python3.11/site-packages/dolfin/function/expression.py:400, in Expression.__init__(self, cpp_code, *args, **kwargs)
    397         if not isinstance(k, str):
    398             raise KeyError("User parameter key must be a string")
--> 400     self._cpp_object = jit.compile_expression(cpp_code, params)
    401     self._parameters = ExpressionParameters(self._cpp_object, params)
    403 if element and degree:

File ~/checkouts/readthedocs.org/user_builds/festim-vv-report/conda/35/lib/python3.11/site-packages/dolfin/function/jit.py:158, in compile_expression(statements, properties)
    154 mpi_comm = properties.pop("mpi_comm", MPI.comm_world)
    155 cpp_data = {'statements': statements, 'properties': properties,
    156             'name': 'expression', 'jit_generate': jit_generate}
--> 158 expression = compile_class(cpp_data, mpi_comm=mpi_comm)
    159 return expression

File ~/checkouts/readthedocs.org/user_builds/festim-vv-report/conda/35/lib/python3.11/site-packages/dolfin/jit/jit.py:170, in compile_class(cpp_data, mpi_comm)
    168     submodule = dijitso.extract_factory_function(module, "create_" + module_name)()
    169 except Exception:
--> 170     raise RuntimeError("Unable to compile C++ code with dijitso")
    172 if name == 'expression':
    173     python_object = cpp.function.make_dolfin_expression(submodule)

RuntimeError: Unable to compile C++ code with dijitso
```

I'm pretty confident #205 will fix this which means you should be able to fix it by adding `gxx` to your environment.

I can reproduce the problem in `quay.io/condaforge/miniforge3` with:

```
mamba create -n dolfin fenics-dolfin mpich matplotlib-base
apt-get -y update; apt-get -y install g++ # install system compiler
# add env to PATH without activating it
export PATH=/opt/conda/envs/dolfin/bin:$PATH
python3 demo_poisson.py
# fails with 
/usr/bin/ld: cannot find /lib64/libpthread.so.0
/usr/bin/ld: cannot find /usr/lib64/libpthread_nonshared.a
collect2: error: ld returned 1 exit status

mamba install -n dolfin gxx
python3 demo_poisson.py  # succeeds
```

This doesn't come up for properly activated environments because the activation scripts set $CXX, so the default `c++` executable is not used.

--- a/recipe/build-dolfin.sh
+++ b/recipe/build-dolfin.sh
@@ -1,8 +1,6 @@
 #!/bin/bash
 set -eux
 
-cd dolfin
-
 # scrub problematic -fdebug-prefix-map from C[XX]FLAGS
 # these are loaded in the clang[++] activate scripts
 export CFLAGS=$(echo $CFLAGS | sed -E 's@\-fdebug\-prefix\-map[^ ]*@@g')

--- a/recipe/build-libdolfin.sh
+++ b/recipe/build-libdolfin.sh
@@ -1,10 +1,6 @@
 #!/bin/bash
 set -eux
 
-unset CMAKE_PREFIX_PATH
-
-cd dolfin
-
 # scrub problematic -fdebug-prefix-map from C[XX]FLAGS
 # these are loaded in the clang[++] activate scripts
 export CFLAGS=$(echo $CFLAGS | sed -E 's@\-fdebug\-prefix\-map[^ ]*@@g')

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -8,7 +8,6 @@ package:
 source:
   - url: https://bitbucket.org/fenics-project/dolfin/downloads/dolfin-{{ version }}.post0.tar.gz
     sha256: 61abdcdb13684ba2a3ba4afb7ea6c7907aa0896a46439d3af7e8848483d4392f
-    folder: dolfin
     patches:
       - boost.patch
       - linuxboost.patch  # [linux]
@@ -151,7 +150,7 @@ outputs:
       commands:
         - bash ${RECIPE_DIR}/parent/test-dolfin.sh
       source_files:
-        - dolfin/python/test
+        - python/test
 
       requires:
         - nose
@@ -180,9 +179,10 @@ outputs:
 about:
   home: http://www.fenicsproject.org
   license: LGPL-3.0-or-later
+  license_family: LGPL
   license_file:
-    - dolfin/COPYING
-    - dolfin/COPYING.LESSER
+    - COPYING
+    - COPYING.LESSER
   summary: 'FEniCS is a collection of free software for automated, efficient solution of differential equations'
 
   description: |

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -17,7 +17,7 @@ source:
       - python-cmake-args.patch
 
 build:
-  number: 47
+  number: 48
   skip: true  # [win]
   # this doesn't actually affect the build hashes
   # so duplicate where the build hash should actually change
@@ -127,6 +127,8 @@ outputs:
         - hdf5 * {{ mpi_prefix }}_*
       run:
         - {{ compiler('cxx') }}
+        # gxx provides default 'c++' executable on linux
+        - gxx  # [linux]
         - python
         # dolfin depends on the boost headers for its own headers, see
         # https://bitbucket.org/fenics-project/dolfin/src/master/dolfin/parameter/Parameters.h#lines-24

--- a/recipe/test-dolfin.sh
+++ b/recipe/test-dolfin.sh
@@ -31,7 +31,7 @@ EOF
 
 python -c 'from dolfin import *; info(parameters["form_compiler"], True)'
 
-pushd "dolfin/python/test/unit"
+pushd "python/test/unit"
 TESTS="jit fem/test_form.py::test_assemble_linear"
 
 RUN_TESTS="python -b -m pytest -vs $TESTS"

